src/utils/reductions.hpp

```diff
@@ -70,16 +70,26 @@ int GetSize(U &v) {
 
 #ifdef MPI_PARALLEL
 // Generic for containers including Kokkos::View and std::vector
-// strips the pointers off of the type
-template <template <typename U, typename... Args> class Container, typename U,
-          typename... Args>
-MPI_Datatype GetType(Container<U, Args...> &v) {
-  return MPITypeMap<typename cleantypes::remove_all_pointers<U>::type>::type();
-}
-template <typename U>
-MPI_Datatype GetType(U &v) {
-  return MPITypeMap<U>::type();
+template <class T, class = typename T::value_type>
+struct value_type_test {
+  using void_t = void;
+};
```

Comment:
Just for clarification, does this now "strip the pointer" and makes it `void` type?
I'm trying to match the functionality of the code to the name. Maybe a brief comment could help.

Reply:
The "`void_t` trick" is some somewhat obscure SFINAE that can be used for checking if a type contains specific other  types or methods. A more clean implementation than what I have in the code is:  
```c++
template <class... Ts> 
using void_t = void; 

template<class T, class = void> 
struct has_foo_method : std::false_type{}; 

template<class T> 
struct has_foo_method<T, void_t<decltype(std::declval<T>().foo())>> : std::true_type {}; 

template<class T, class = void> 
struct has_value_type : std::false_type{}; 

template<class T> 
struct has_value_type<T, void_t<typename T::value_type>> : std::true_type {}; 

struct A {};
struct B {using value_type = int; void foo(){}};

int main() {
    std::cout << has_value_type<A>::value << std::endl;
    std::cout << has_value_type<B>::value << std::endl;
    std::cout << has_foo_method<A>::value << std::endl;
    std::cout << has_foo_method<B>::value << std::endl;
}
``` 
The type doesn't necessarily need to be `void` (for instance there are actually some reasons why it might be a little better to use `bool`). What is happening is `void_t<...> = void` for any well formed `...`, so the compiler chooses the specialization (which inherits from `std::true_type`) as the best match since the second template parameter defaults to `void`. If `...` is not well formed (i.e. the class does not have the requested method) the specialization fails to compile and the general template (which inherits from `std::false_type`) is chosen. 

I noticed last night that there are a few other functions above this one that need to be fixed as well, I will try to go through later this weekend and clean everything up. I will also add a little explanation because I agree this is a bit inscrutable.

Reply:
After thinking about this and looking around a little yesterday, I think I figured out how to implement this in a way that is a little bit more general and doesn't look so obscure. The suggested implementation is in #694.

Reply:
done.